src/helpers/setUpEdgeFunction.js

```diff
@@ -0,0 +1,79 @@
+const { existsSync, readdirSync } = require('node:fs')
+const { writeFile, mkdir, readFile } = require('node:fs/promises')
+const { join, relative } = require('node:path')
+
+const { readJson } = require('fs-extra')
+
+/**
+ * Recursively lists all files in a directory.
+ */
+const getAllFilesIn = (dir) =>
+  readdirSync(dir, { withFileTypes: true }).flatMap((dirent) => {
+    if (dirent.isDirectory()) {
+      return getAllFilesIn(join(dir, dirent.name)).map((path) => join(dirent.name, path))
+    }
+    return [join(dir, dirent.name)]
+  })
+
+const setUpEdgeFunction = async ({ angularJson, projectName, netlifyConfig, constants, failBuild }) => {
+  const project = angularJson.projects[projectName]
+  const {
+    architect: { build },
+  } = project
+  const outputDir = build?.options?.outputPath
+  if (!outputDir || !existsSync(outputDir)) {
+    return failBuild('Could not find build output directory')
+  }
+
+  console.log(`Writing Angular SSR Edge Function for project "${projectName}" ...`)
+
+  netlifyConfig.build.publish = join(outputDir, 'browser')
+
+  const edgeFunctionDir = join(constants.INTERNAL_EDGE_FUNCTIONS_SRC, 'angular-ssr')
+  await mkdir(edgeFunctionDir, { recursive: true })
+
+  const serverDistRoot = join(outputDir, 'server')
+  const html = await readFile(join(serverDistRoot, 'index.server.html'), 'utf-8')
+  const staticFiles = getAllFilesIn(join(outputDir, 'browser')).map(
+    (path) => `/${relative(join(outputDir, 'browser'), path)}`,
+  )
+
+  const { routes: prerenderedRoutes } = await readJson(join(outputDir, 'prerendered-routes.json'))
+  const excludedPaths = [...staticFiles, ...prerenderedRoutes]
+
+  const polyfills = `
+  import process from "node:process"
+
+  globalThis.process = process
+  globalThis.global = globalThis
```

Comment:
Any reason this needs to be in a separate file? It's pretty small

Reply:
Yep, needs to be in a separate file to ensure it's initialised first. ESM spec says that modules are initialised leaf-first, root-last, and children are initialised in order of source code appearance. So by placing this in a separate (leaf) module + importing it first, we ensure this code runs first, no matter what.

Reply:
Yep, needs to be in a separate file to ensure it's initialised first. ESM spec says that modules are initialised leaf-first, root-last, and children are initialised in order of source code appearance. So by placing this in a separate (leaf) module + importing it first, we ensure this code runs first, no matter what.

Reply:
Makes sense. Could you add a comment to that effect?

Reply:
done in https://github.com/netlify/netlify-plugin-angular-universal/pull/67/commits/01892694c332934b622b143a8079e843a36168c7